src/lib/stats.rs

```diff
@@ -15,14 +22,80 @@ pub fn pearson_correlation_coefficient(xs: &Vec<f32>, ys: &Vec<f32>) -> f64{
 
     let lval: f64 = 1.0 / ((n as f64) - 1.0);
 
-    return rval * lval;
+    let final_val = rval * lval;
+
+    if final_val.is_nan() {
+        return 0.0;
+    }
+
+    return final_val;
 }
 
 pub fn pairinate(a: &BinaryTimeSeries, b: &BinaryTimeSeries)
-                 -> (Vec<f32>, Vec<f32>){
-    if a.data.len() != b.data.len(){
-        panic!("AHHHH! Everything is broken. Those datas don't match.");
+                 -> Option<(Vec<f32>, Vec<f32>)>{
+
+    let a_start = try_opt!(get_start_index(a, b));
+    let b_start = try_opt!(get_start_index(b, a));
+    let a_end = try_opt!(get_end_index(a, b));
+    let b_end = try_opt!(get_end_index(b, a));
+
+    if (a_end - a_start) >= (b_end - b_start) {
+        // Only one point of overlap, cannot correlate
+        if a_end - a_start == 0 {return None};
+        return Some(interpolate(&a.data[a_start .. a_end], &b.data[..]));
+    }else{
+        return Some(interpolate(&b.data[b_start .. b_end], &a.data[..]));
+    }
+}
+
+fn get_start_index(series: &BinaryTimeSeries, target: &BinaryTimeSeries) -> Option<usize>{
+    let mut start_index = 0;
+    while series.data[start_index].t < target.data[0].t{
+        start_index += 1;
+        if start_index >= series.data.len() {
+            return None;
+        }
+    }
+    return Some(start_index);
+}
+
+fn get_end_index(series: &BinaryTimeSeries, target: &BinaryTimeSeries) -> Option<usize>{
+    let mut end_index = series.data.len() - 1;
+
+    while series.data[end_index].t < target.data[target.data.len() - 1].t{
+        if end_index == 0{
+            return None;
+        }
+        end_index -= 1;
+    }
+    return Some(end_index)
+}
+
+fn interpolate(base: &[Point], other: &[Point]) -> (Vec<f32>, Vec<f32>){
+    let mut search_index = 0;
+
+    let mut base_data = Vec::<f32>::new();
+    let mut other_data = Vec::<f32>::new();
+
+    for data_point in base {
+        base_data.push(data_point.val);
+        while other[search_index + 1].t <= data_point.t{
```

Comment:
I'm having trouble proving to myself that this can't index out of bounds, but I'm pretty sure it is right and the kind of subset-ish overlap property means it never does.